### ISSUE TITLE
Address matching -- normalize directional (N,S,E,W) components of an address

### ISSUE THREAD

**Issue by RDmitchell** [SEED-platform/seed#315]
The program still has a problem with false positives when only matching on address. It seems that part of the problem is we are not dealing with enough of the possible address "suffixes", such as the directionals (S, South). I will put together an example data set for testing this.

**Comment by RDmitchell** [comment]
There are two example files in the GDrive issue folder for this issue:

https://drive.google.com/a/lbl.gov/folderview?id=0B3fTKpZ9Dx7LcDU2Y3lJVGpiV0U&usp=sharing

Here is a screenshot of the data in the two files, and how the program should match them up if they are just matched on Address, which is the only field that is common between them. The addresses are exactly the same except for the "South" suffix that is added to one of them, and which indicates a completely separate address. We are not dealing with those directional suffixes (or possibly any other type of suffix after the street type, ie, Ave, St, etc).

![image](https://cloud.githubusercontent.com/assets/6314950/9013975/3aca8600-3774-11e5-93b2-0eed2b647867.png)

In SEED, first import the Building Data.xls file, map it, and save the mapping (which causes the matching step).

![image](https://cloud.githubusercontent.com/assets/6314950/9014037/a14ef4d8-3774-11e5-95c2-4fc964a578b2.png)

Here is the mapping for the Building Data
![image](https://cloud.githubusercontent.com/assets/6314950/9014077/e74818e8-3774-11e5-8612-2e2db893103f.png)

Program makes 2 master building records, one for "Willow Avenue" and one for "Willow Avenue South". 
![image](https://cloud.githubusercontent.com/assets/6314950/9014326/97fc0ffe-3776-11e5-9d47-70e7bd975646.png)
Building Detail view for Willow Avenue
![image](https://cloud.githubusercontent.com/assets/6314950/9014358/c3201ca2-3776-11e5-962c-4d26de144901.png)
Building Detail view for Willow Avenue South
![image](https://cloud.githubusercontent.com/assets/6314950/9014385/e9aebbf8-3776-11e5-884d-e16dee9f1c7a.png)

Next import the "PM data.csv" file, choosing the Portfolio Manager button on the Uploading Your building data screen
![image](https://cloud.githubusercontent.com/assets/6314950/9014410/11959d94-3777-11e5-9fa2-2c708b9923c2.png)

Here is the mapping for the PM file
![image](https://cloud.githubusercontent.com/assets/6314950/9014477/74216e66-3777-11e5-999a-b2eee8721ade.png)

Map, save mappings, and the program will match the records. Go to the matching screen and you will see the following:
![image](https://cloud.githubusercontent.com/assets/6314950/9014522/be63062e-3777-11e5-9b2b-b706f351fcf3.png)

The program has matched two PM records, one with address of Willow Avenue South and one with Willow Avenue, which is wrong. In addition, it has matched a building record with address of Willow Avenue with the PM record of Willow Avenue South. 

![image](https://cloud.githubusercontent.com/assets/6314950/9015241/7106a46c-377c-11e5-900f-b3a1648cb7f3.png)
The program should have matched the PM and Building records with the address "Willow Avenue" and the PM and Building records with the address "Willow Avenue South".

![image](https://cloud.githubusercontent.com/assets/6314950/9015307/e33bb810-377c-11e5-9b15-216a44da6274.png)

Looking at the Building list, there are now two master building records both with the address "Willow Avenue South", one that has 2 PM records and one Building record incorrectly associated with it, and one that has just a Building Record.
![image](https://cloud.githubusercontent.com/assets/6314950/9015436/cbbaab14-377d-11e5-967e-974b726a14db.png)

Here is the Building Detail view for the first Master record, that has the 3 associated records with it.
![image](https://cloud.githubusercontent.com/assets/6314950/9015548/89401cf0-377e-11e5-898d-9670b4959d98.png)

Here is the Master record that the program should have matched to the PM record with the address of "Willow Avenue South". But this record doesn't have any PM records associated with it. 
![image](https://cloud.githubusercontent.com/assets/6314950/9015574/b46d0596-377e-11e5-9ce7-bdbc481cac98.png)

**Comment by cycustodio** [comment]
The normalize address function I believe only captures the street number, street name, and the street type/suffix (e.g. "123 Test St."). The library itself has the capability to get the last parts of the address, apartment number and the directions ("south", "NE", etc.), but wasn't included in the normalize function. 

What happens is when we try to match the database values to the currently uploaded values "111 Willow Avenue South" turns into "111 Willow Avenue".

Here is the part where the normalize address function captures the parts of the address. 
```
 if 'house' in addr and addr['house'] is not None:
        normalized_address = addr['house'].lstrip("0")  # some addresses have leading zeros, strip them here

    if 'street_name' in addr and addr['street_name'] is not None:
        normalized_address = normalized_address + ' ' + addr['street_name']

    if 'street_type' in addr and addr['street_type'] is not None:
        normalized_address = normalized_address + ' ' + addr['street_type']
```

For reference see Issue (https://github.com/SEED-platform/seed/issues/210)

**Comment by dnfehren-elevate** [comment]
Perhaps beyond the scope of this specific issue but the address parsing module also has some issues with address ranges. If there is no whitespace in an address range it parses the 'house' appropriately (though not necessarily in a useful way - there is no indication of range start and range end). If there is whitespace then the parser is less predictable and the output is not so well formed. 

Address ranges and matching datasets on their permutations have been a real struggle for me so this might add unwanted complexity to your normalization task but getting it right could improve matching.

example test addresses

    322 S Green St #house:322
    300-22 S Green St #house:300-22
    300/322 S Green St #house:300/322
    300 322 S Green St #house:300,street_name:'322 S green'

**Comment by RDmitchell** [comment]
Daniel -- we definitely want to improve the address matching over time, and this is another good case that we need to tackle.

I am going to make this a separate issue and basically copy your comment into it, to deal with address ranges. 

Thanks for your feedback !

**Comment by RDmitchell** [comment]
Nick -- can you evaluate what the level of effort would be to add the capability to normalize addresses on at least direction (N, North, NE, etc) and possibly Apt #.

See Claudine's comment from July 31 that the library we are using to do the address normalization can deal with apartment number and directions, but we aren't using that capability. We need to add that functionality.

I am assigning it a milestone of Sept 2015, but if it is not complete by our PR deadline (something like 9/18) that is ok, it can go into the next release after that. Thx

**Comment by RDmitchell** [comment]
I can put together some example files from real city data (after anonymizing it) -- lots of cities have identical addresses except for the directions, ie, 1 Main Street NE vs 1 Main Street SW, which we need to be able to distinguish between. Salt Lake City and DC in particular have their street pattern set up on a grid that depends on the direction suffix to differentiate addresses.

**Comment by RDmitchell** [comment]
The Building data.xlsx file in the Issue 315 folder has a few examples of how addresses can be represented with directions. 

https://drive.google.com/a/lbl.gov/folderview?id=0B3fTKpZ9Dx7LcDU2Y3lJVGpiV0U&usp=sharing

**Comment by nicholasserra** [comment]
Ok, so to be clear, the issue is that the matching code is stripping directions, which is leading to matches that really shouldn't be matched?

**Comment by RDmitchell** [comment]
Correct. And I believe that the library we are using to do address normalization includes the capability to include the directions in the address normalization so that 
100 Main South would match to 100 Main S, but not to 100 Main

**Comment by nicholasserra** [comment]
Looks like the parser we're using can format directions, but doesn't parse them AS directions to be easily pulled out. There's also an idea of trailing vs preceding directions that we might have to worry about. Aka `528 S. Hazelwood` vs `528 Hazelwood South`.

Can we get a list going of common directions we might see? I think the best course of action is to try to get some address handling merged upstream into that library, as it's on github.

**Comment by RDmitchell** [comment]
The sample file that I reference above starts to do this, but we can be more precise about what we want to deal with.

This is probably what we need to cover (and maybe there are more subtle variations of this which I will look for):

<street Number>  <direction>  <street name>  <direction> 

* where street name can be multiple words
* where direction can be:
 * N
 * North
 * NE
 * NW
 * S
 * South
 * SE
 * SW
 * E
 * East
 * W
 * West

**Comment by RDmitchell** [comment]
Examples from file

![image](https://cloud.githubusercontent.com/assets/6314950/9644659/326dee6c-517c-11e5-93c1-2a0c815d9514.png)

**Comment by RDmitchell** [comment]
Nick -- does this help? Do you need more info? thx robin

**Comment by nicholasserra** [comment]
This is perfect. I'm going to fork that library and add a bunch of tests around these addresses and try to get the parser working with them.

**Comment by RDmitchell** [comment]
great !

**Comment by rebrownlbnl** [comment]
I know we went through this when we first added address parsing, but the parsing library we are using is not at all active and has never had a pull request (but it does seem to do most of what we want). Maybe it's only relevant when we have time to make a big change, but this seems like a much better supported package that, at first glance, does similar things: https://github.com/datamade/usaddress

**Comment by nicholasserra** [comment]
That library looks nice. I think it would be easier to switch to that than to update the other one.

**Comment by nicholasserra** [comment]
So I have something working pretty nicely using `usaddress` library as the parser and using the existing `streetaddress` library as the formatter. If everyone's okay with using both dependencies I can probably have a PR in to handle this tomorrow.

**Comment by RDmitchell** [comment]
Sounds good. We can test it out once your PR is in and merged. Thanks !

**Comment by rebrownlbnl** [comment]
I think it's okay to have both dependencies. Did you look at the test cases that @dnfehren-elevate added above in this thread? Parsing address ranges sounds useful.

**Comment by RDmitchell** [comment]
But let's get the direction stuff figured out first before diving into address ranges? Unless for some reason it makes sense to do it all now. I'd be tempted to get one figured out and tested, and then tackle the range issue. 

Nick, let me know what makes the most sense to you. If you would prefer to deal with both issues now, because you are diving into it, that's fine, but I am also fine with doing each as a separate issue / task.

**Comment by nicholasserra** [comment]
Let me get this direction stuff in first. I'm setting it up so that test cases for specific addresses will be easier to implement too.

**Comment by rebrownlbnl** [comment]
agree with fixing the direction stuff first.

**Comment by nicholasserra** [comment]
So for direction normalization, i'm going to shorten all directions using this map:

```
    direction_map = {
        'east' : 'e',
        'west' : 'w', 
        'north' : 'n',
        'south' : 's',
        'northeast': 'ne',
        'northwest': 'nw',
        'southeast': 'se',
        'southwest': 'sw'
    }
```

So both `100 Main South` and `100 Main S` will normalize to `100 main s`.

**Comment by nicholasserra** [comment]
Here's tests i'm adding and their output:
`input -> output`

```
'123 Test St. NE' ->  '123 test st ne'
'123 South Test St.' ->  '123 s test st'
'123 Test St. Northeast' ->  '123 test st ne'
'111 S West Main' ->  '111 s west main'
'555 11th St. NW' ->  '555 11th st nw'
'100 Main S' ->  '100 main s'
'100 Main South' ->  '100 main s'
```

**Comment by nicholasserra** [comment]
PR in, assigning back.

**Comment by RDmitchell** [comment]
Claudine -- I am assigning this to you for testing. Thx. Robin

**Comment by cycustodio** [comment]
The files used in this test can be found here:
https://drive.google.com/a/lbl.gov/folderview?id=0B3fTKpZ9Dx7LcDU2Y3lJVGpiV0U&usp=sharing
Used covered-buildings-sample and portfolio-manager-sample_addresstest. These are modifications to a subset of the sample files which only match on ```Address Line 1```. 

The directions are getting parsed and the other weird address things (leading zeros, abbreviations of street types, etc.) are getting normalized and matched. The differences between the addresses are highlighted. 
![address_matching_wdirections](https://cloud.githubusercontent.com/assets/8461331/9801340/51ba71cc-57c7-11e5-849e-7100d021a7d6.png)

```Cherry Loop``` vs ```W Cherry Loop``` example is the case similar to the one that started the thread. These addresses don't match anymore. The one case I can't figure out is "Peach Ave. East" vs "Peach Avenue E".
![address_unmatched](https://cloud.githubusercontent.com/assets/8461331/9801344/53afc1b2-57c7-11e5-8f4e-462c348d4a72.png)

**Comment by nicholasserra** [comment]
Sooo did it work or not work? hah :)

**Comment by nicholasserra** [comment]
I'm understanding this as it worked for some cases, and we need to handle more cases

**Comment by RDmitchell** [comment]
Claudine -- you need to be specific about what cases it is not handling so that Nick can work on those.

**Comment by cycustodio** [comment]
I messed up the color of the highlight boxes, the image should be fixed now. Green means the matches/non- matches are valid.

**Comment by cycustodio** [comment]
TL;DR It's working except for the corner case that is 
Peach Ave. East -- abbreviated street name with punctuation, direction at the end

**Comment by nicholasserra** [comment]
Nice!

**Comment by RDmitchell** [comment]
Claudine, can you make a new issue for the case not covered, per your previous comment? Thx Robin
I will close this issue
